tests/core/data/test_tensordense.py

```diff
@@ -0,0 +1,234 @@
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert_almost_equal(test.to_array(), list_dense)
+
+    @pytest.mark.parametrize('dtype', ['complex128',
+                                       'float64',
+                                       'int32', 'int64',
+                                       'uint32'])
+    def test_init_from_ndarray_other_dtype(self, shape, dtype):
+        numpy_dense = np.random.rand(*shape).astype(dtype, casting='unsafe')
+        test = DenseTensor(numpy_dense)
+        assert test.shape == shape
+        assert test._tf.dtype == tf.complex128
+        assert test._tf.shape == shape
+        assert np.all(test.to_array() == numpy_dense)
+
+    @pytest.mark.parametrize('dtype', ['complex128',
+                                       'float64',
+                                       'int32', 'int64',
+                                       'uint32'])
+    def test_init_from_tensor_other_dtype(self, shape, dtype):
+        numpy_dense = np.random.rand(*shape).astype(dtype, casting='unsafe')
+        tensor = tf.constant(numpy_dense)
+        test = DenseTensor(tensor)
+        assert test.shape == shape
+        assert test._tf.shape == shape
+        assert test._tf.dtype == tf.complex128
+
+        tensor = tf.cast(tensor, dtype=tf.complex128)
+        assert np.all(test._tf == tensor)
+
+    @pytest.mark.parametrize(['arg', 'kwargs', 'error'], [
+        pytest.param(_valid_tensor(), {'shape': ()}, ValueError,
+                     id="numpy-shape 0 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (1,)}, ValueError,
+                     id="numpy-shape 1 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (None, None)}, ValueError,
+                     id="numpy-shape None tuple"),
+        pytest.param(_valid_tensor(), {'shape': [2, 2]}, ValueError,
+                     id="numpy-shape list"),
+        pytest.param(_valid_tensor(), {'shape': (1, 2, 3)}, ValueError,
+                     id="numpy-shape 3 tuple"),
+        pytest.param(_valid_tensor(), {'shape': (-1, 1)}, ValueError,
+                     id="numpy-negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (-4, -4)}, ValueError,
+                     id="numpy-both negative shape"),
+        pytest.param(_valid_tensor(), {'shape': (1213, 1217)}, ValueError,
+                     id="numpy-different shape"),
+    ])
+    def test_init_from_wrong_input(self, arg, kwargs, error):
+        """
+        Test that the __init__ method raises a suitable error when passed
+        incorrectly formatted inputs.
+
+        This test also serves as a *partial* check that Dense safely handles
+        deallocation in the presence of exceptions in its __init__ method.  If
+        the tests segfault, it's quite likely that the memory management isn't
+        being done correctly in the hand-off us setting our data buffers up and
+        marking the numpy actually owns the data.
```

Comment:
Please don't just copy-and-paste code - incorrect documentation can even be worse than no documentation at all.

Reply:
Sorry for this. I will go through all the comments in the test and ensure they are correct.